exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/DrillConstExecutor.java

```diff
@@ -143,116 +156,176 @@ public void reduce(RexBuilder rexBuilder, List<RexNode> constExps, List<RexNode>
         continue;
       }
 
-        switch(materializedExpr.getMajorType().getMinorType()) {
-          case INT:
-            reducedValues.add(rexBuilder.makeLiteral(
-                new BigDecimal(((IntHolder)output).value),
-                TypeInferenceUtils.createCalciteTypeWithNullability(typeFactory, SqlTypeName.INTEGER, newCall.getType().isNullable()),
-                false));
-            break;
-          case BIGINT:
-            reducedValues.add(rexBuilder.makeLiteral(
-                new BigDecimal(((BigIntHolder)output).value),
-                TypeInferenceUtils.createCalciteTypeWithNullability(typeFactory, SqlTypeName.BIGINT, newCall.getType().isNullable()),
-                false));
-            break;
-          case FLOAT4:
-            reducedValues.add(rexBuilder.makeLiteral(
-                new BigDecimal(((Float4Holder)output).value),
-                TypeInferenceUtils.createCalciteTypeWithNullability(typeFactory, SqlTypeName.FLOAT, newCall.getType().isNullable()),
-                false));
-            break;
-          case FLOAT8:
-            reducedValues.add(rexBuilder.makeLiteral(
-                new BigDecimal(((Float8Holder)output).value),
-                TypeInferenceUtils.createCalciteTypeWithNullability(typeFactory, SqlTypeName.DOUBLE, newCall.getType().isNullable()),
-                false));
-            break;
-          case VARCHAR:
-            reducedValues.add(rexBuilder.makeCharLiteral(
-                new NlsString(StringFunctionHelpers.getStringFromVarCharHolder((VarCharHolder)output), null, null)));
-            break;
-          case BIT:
-            reducedValues.add(rexBuilder.makeLiteral(
-                ((BitHolder)output).value == 1 ? true : false,
-                TypeInferenceUtils.createCalciteTypeWithNullability(typeFactory, SqlTypeName.BOOLEAN, newCall.getType().isNullable()),
-                false));
-            break;
-          case DATE:
-            reducedValues.add(rexBuilder.makeLiteral(
-                new DateTime(((DateHolder) output).value, DateTimeZone.UTC).toCalendar(null),
-                TypeInferenceUtils.createCalciteTypeWithNullability(typeFactory, SqlTypeName.DATE, newCall.getType().isNullable()),
-                false));
-            break;
-          case DECIMAL9:
-            reducedValues.add(rexBuilder.makeLiteral(
-                new BigDecimal(BigInteger.valueOf(((Decimal9Holder) output).value), ((Decimal9Holder)output).scale),
+      Function<ValueHolder, RexNode> literator = new Function<ValueHolder, RexNode>() {
```

Comment:
I tried to minimize code under the switch cases. As you can see, before my pull request "reducedValues.add()" calls under any of cases. 
At first I tried to move switch to private method, somethig like "private RexNode getRexNode()" but there was a lot of arguments to pass - typeFactory, builder, etc, etc. Unfortunately we can't use Java8 lambdas, so the function interface(de facto it's nested function) is the optimal solution I think.